### ISSUE TITLE
Large export int var is unpredictable when modified from the editor

### ISSUE THREAD

**Issue by Netforza** [godotengine/godot#33095]
**Godot version:** 
v3.1.1.stable.official

**OS/device including version:**
Ubuntu 18.04.3 LTS 64-bit
Intel® Core™ i7-3930K CPU @ 3.20GHz × 12
GeForce GTX 1080 Ti/PCIe/SSE2
32GB ram

**Issue description:**
Large exported INT variables
`export var a: int = 10000000000000`
**If changed manually in the editor by typing or deleting a digit will become an unpredictable number.**
If the number is edited to a float value that has a dot, it will round to a predictable int number
When using the input arrows to modify the number incrementally everything seems to work fine.
When dragging the mouse to modify the number incrementally it also seems to work fine.

**Steps to reproduce:**
add to a script: 
`export var a: int = 10000000000000`
In the Script Variables section in the editor, delete the last digit and press enter.

**Comment by qarmin** [comment]
Try to use 3.2 alpha, because since #31513, Godot should show error, and clamp value to range (MIN_INT, MAX_INT)

**Comment by Netforza** [comment]
I checked in  3.2 alpha2
The behavior is the same but it throws this error in console:
**Cannot represent 10000000000000 as integer, provided value is too big.**

I should mention that this value is smaller than max 64bit integer that is supposedly used in Godot.
And modifying the value by using the arrows in the input as well as mouse dragging works as expected.
Plus the actual int var if not modified in the editor works as expected in the script so the reason for clumping it in the editor when editing via typing is not clear.

**Comment by CastilloDel** [comment]
In Godot 3.2 I get 2147483647 which is equal to 2^31-1 which is the maximum value of a 64-bit integer. It's true that you can still add with the arrows but if you try to write a value greater than 2147483647 it will throw the error and clamp it.
I guess the error is the fact that you are still able to add value using the arrows, but I find odd that it works as expected in the script

**Comment by pycbouh** [comment]
No, the issue is that Godot Engine supports 64-bit signed ints, but Inspector does not. Values above the maximum 32-bit signed int work within some range, but then fail to be handled properly.

In fact, you can set it via code to the maximum 64-bit signed value, but in inspector it would show up as **9223372036854775800**, which is 7 units shy, and won't allow any changes with arrows. Dragging in the input field produces some changes to the value, which looks like this in the Inspector:
* 9223372036854775800
* 9223372036854774800
* 9223372036854773800
* 9223372036854772700
* 9223372036854771700
* etc

If you set a setter function and print the actual value, then you can see that it is more precise:
* 9223372036854775807
* 9223372036854774784
* 9223372036854773760
* 9223372036854772736
* 9223372036854771712
* etc

The problems don't stop in the Inspector, though. Inspector uses something that looks like a `SpinBox` control. And indeed, you can setup a `SpinBox` in your UI scene and get similar issues. You have to set `max_value` property in code, because Inspector won't allow to set it to a value above **2147483647**. After that you can run the scene and try to paste into the field the value of **9223372036854775807**. This will cause an error
```
E 0:00:03.261   to_int: Cannot represent 9223372036854775807 as integer, provided value is too big.
  <C++ Error>   Condition "integer > 2147483647 / 10" is true. Returned: sign == 1 ? 2147483647 : (-2147483647 - 1)
  <C++ Source>  core/ustring.cpp:1792 @ to_int()
```

Actually, you don't need to change default `max_value` to have this error. Even if it sits at 100, there will be an error in debug. And it starts as low as **2147483650**, which is 3 units more than the maximum 32-bit signed integer.
```
E 0:01:47.811   to_int: Cannot represent 2147483650 as integer, provided value is too big.
  <C++ Error>   Condition "integer > 2147483647 / 10" is true. Returned: sign == 1 ? 2147483647 : (-2147483647 - 1)
  <C++ Source>  core/ustring.cpp:1792 @ to_int()
```

**Comment by pycbouh** [comment]
Problem is still present in master branch. And it looks like there are two distinct problems: with export settings and with the inspector controls.

### Inspector precision issue
If `export` is used without any additional hints, then we have a precision problem in inspector control (or `SpinBox`/`Range` in general). Somewhere between **999999999999999** and **9999999999999999** (as far as I can tell) values start to lose precision, begin skipping by 2 and then by more. Until we rich the maximum int64 value of **9223372036854775807**, which cannot be reached due to this precision issue. So value gets bigger still and overflows into **-9223372036854775808**. Next 3 values cause no error and silently overflow as well. Error starts to appear only with **9223372036854775810**.

Overflow starts to happen at **9223372036854775296**, which is 513 units away from the largest value, that does not err. Other interesting numbers are **9223372036854774273** and **9223372036854775295**, which define a range, that all rounds to **9223372036854774784**. That's 1023 values. Previous range is from **9223372036854773248** to **9223372036854774272**, rounding at **9223372036854773248**. That's 1025 values. So while the precision issue is growing with higher numbers, it seems to be fluctuating a bit.

It seems, something in master was changed after all, as the error message now tells about 64-bit ints instead of 32-bit ints:
>  Cannot represent 9223372036854775810 as 64-bit integer, provided value is too big.

So it seems to now rely on `String::to_int64` instead of `String::to_int`, like it did before. This much I can tell, however I was unable to find the source of the precision issues.


### Export hint max int issue
If `export` is used with hints, defining an allowed range, then there is a separate issue with 32- and 64-bit limits. This is a correct statement, that allows any int within int64 range:
```
export(int, -9223372036854775808, 9223372036854775807)
```

It does not work. And neither does
```
export(int, 9223372036854775807)
```

It seems, that these values cannot be parsed correctly. In the first case any value entered in inspector will turn into the max int32 value of **2147483647**, even if the value entered is within int32 range. Just any value will turn into **2147483647** with no exception. In the second case, values between **0** and **2147483647** are allowed, as the left bracket can be parsed.

In each case there is an error for each big value within export hints (even in the second case, where the value coded is **9223372036854775807**):
> Cannot represent 9223372036854775808 as integer, provided value is too big.

-----

This is what happens when an int is exported, but floats have their own problems. I'm not sure, how big a float can be in Godot, but it seems that only the first six digits are stored for values in inspector. Values can still go up to the max int64, but every digit after the sixth is going to be a 0:
```
2147483647
2147480000
```
```
9223372036854775807
9223370000000000000
```

Interestingly, value does not change in the input field right away. Rather it is stored incorrectly, and the field is reset to an incorrect value after a scene is reloaded in editor. The issue is immediately visible if the scene is run as well.

-----

All this is very frustrating, as various issues just crop up on you from many directions in a feature that is supposed to just work. And I think this is a pretty serious set of issues, as a lot of it can be hidden from a developer until someone runs into it and figures out, what is going on.

**Comment by aaronfranke** [comment]
@pycbouh About the display not working properly: as of #35187, if you try to use an integer larger than 2^53 that doesn't perfectly map to a float, Godot will show this warning:

> `"Cannot reliably represent '" + itos(val) + "' in the inspector, value is too large."`
